Status bar icon toggle

tomglynch opened the pull request:
Hi @lostintangent, hope you're doing well. GitDoc is a great tool, but I wanted a feature to have it always show the status bar icon rather than it disappearing when disabled (#89 ). In fixing that, I also ended up fixing #90 as it affected the icons being shown. I handled icons for matching the File Patterns too. Have a look and let me know what you think?

Cheers, 

Tom

tomglynch committed:
change status bar icon to be a toggle that always shows

tomglynch committed:
made a different icon show for when the file pattern does not match

tomglynch committed:
fixed GitDoc inactive on codespace start, regardless of being enabled in setting issue

tomglynch committed:
fixed mirror with circle slash showing momentarily for each change of file being viewed

tomglynch committed:
update readme

tomglynch commented:
@lostintangent just a follow up comment to see if you have any interest in merging these changes? Cheers

awalland commented:
i would also love seeing this merged, since i have the same problem described in #90

tomglynch commented:
@lostintangent if no response soon I think I'll pop the fork up on the vs code extension store.

lostintangent commented:
@tomglynch hey apologies! I'll review this tomorrow and we can get it merged this week

lostintangent commented on src/extension.ts:
```diff
@@ -25,9 +48,38 @@ export async function activate(context: vscode.ExtensionContext) {
+    (enabled) => {
+      updateContext(enabled, true);
+      checkEnabled(git);
+      debouncedUpdateStatusBar(vscode.window.activeTextEditor);
```
I think this should be handled by the checkEnabled function, and effectively revert to the previous code. That way this reaction is simply watching for enabled changes and then updating the UX based on that (via checkEnabled)

lostintangent replied:
Actually, you already have a reaction in the watcher module that will update the status bar when the store.enabled changes. So I think we can just remove this?

lostintangent commented on src/extension.ts:
```diff
@@ -25,9 +48,38 @@ export async function activate(context: vscode.ExtensionContext) {
 
+  // Create a debounced version of updateStatusBarItem
+  let updateTimeout: NodeJS.Timeout | null = null;
+  const debouncedUpdateStatusBar = (editor: vscode.TextEditor | undefined) => {
```
Do we actually need this denounced handler? I don't imagine that a user could change config settings, or their active file, faster than 50ms. At which point, this might be an unnecessary guardrail?

lostintangent commented on src/extension.ts:
```diff
@@ -37,6 +89,7 @@ export async function activate(context: vscode.ExtensionContext) {
         e.affectsConfiguration("gitdoc.autoCommitDelay") ||
         e.affectsConfiguration("gitdoc.filePattern")) {
         checkEnabled(git);
+        debouncedUpdateStatusBar(vscode.window.activeTextEditor);
```
Similar to my comment above, I think we should update the status bar in the checkEnabled function, and simplify this code.

lostintangent reviewed (commented):
Thanks for this! My comments below are essentially just thoughts on how to simplify the implementation, since I think we may be duplicating logic that we don't need. But I'm out of town and haven't been able to actually run/debug the code. So I'm just thinking out loud 👍

Additionally, since this extension has millions of installs, and this PR is making a notable UX change, we should add a setting to control whether to always show the status bar or not. And default it to false. That way the current behavior doesn't change, but you and others can opt into having the status bar always visible.

tomglynch committed:
Fix status bar icon visibility and codespace initialization issues

tomglynch commented:
Thanks for all your comments @lostintangent! I agree with everything and have applied them to my latest commit.
1. Added the `gitdoc.alwaysShowStatusBarIcon` setting (defaulting to false) as you suggested
2. Moved the `matches` function to watcher.ts and made it exported for reuse
3. Simplified the initialization code to use `config.enabled` directly
4. Removed redundant context updates since `checkEnabled` handles this
5. Removed the debounced handler as you suggested it was unnecessary
6. Consolidated status bar updates into the `checkEnabled` function

The changes maintain backward compatibility while allowing users to opt into the new behavior. Let me know if you'd like me to make any additional adjustments. 

I also added a few tests that I haven't pushed yet if you'd like to take a look - but perhaps that's a seperate PR.

lgruen commented:
@lostintangent Thanks for your work on this! Any chance to get this PR merged? I think a couple of people are affected by #90.

tomglynch commented:
@lostintangent

ascholerChemeketa commented:
Would love to see this merged. Currently resorting to a fork to use this in courses to avoid https://github.com/lostintangent/gitdoc/issues/90.

owenarden commented:
I am also affected by this. +1 for merging :)
